src/hotspot/share/gc/shenandoah/shenandoahEvacOOMHandler.hpp

```diff
@@ -59,54 +59,158 @@ class ShenandoahEvacOOMCounter {
+ *     resolution of lingering forwarding pointers that may be scattered throughout
+ *     the previously selected cset when evacuation is abandoned before it has been
+ *     completed.
+ *
+ *  2. If any thread A fails to evacuate object X, it will wait to see if some
+ *     other mutator or GC worker thread can successfully evacuate object X.  At the moment
+ *     thread A fails to allocate, it launches the OOM-during-evacuation protocol.  There
+ *     is no going back (even though some other thread may successfully evacuate object X).
+ *
+ *  3. The protocol consists of:
+ *
+ *     a) Thread A sets internal state to indicate that OOM-during-evac has been
+ *        encountered.
+ *     b) Thread A now waits for all other threads to finish any ongoing allocations
+ *        for evacuation that might be in process.
+ *     c) Other threads that subsequently announce intent to allocate for evacuation are
+ *        informed that the OOM-during-evac protocol has been initated.  As with thread A,
+ *        these threads also wait for all other threads to finish any ongoing allocations
+ *        for evacuation that might be in process.
+ *     d) After all threads have finished whatever allocations for evacuation they
+ *        were in the process of performing, the evacution state of the heap is considered
+ *        to be "frozen".  At this point, some cset objects may have been successfully
+ *        evacuated and some cset objects may have failed to evacuate.  There will be
+ *        no more evaucations until we STW and perform Full GC.
+ *     e) Now, all of the threads that were waiting for evacuating threads to finish
+ *        allocations that were in progress are allowed to run, but they are not allowed
+ *        to allocate for evacuation.  Additional threads that newly announce intent to
+ *        allocate for evacuation are immediately allowed to continue running, also without
+ *        authorization to allocate.
+ *     f) Threads that desire to allocate for evacuation but are not authorized to do so
+ *        simply consult the head of each cset object.  If the header denotes that the
+ *        object has been evacuated by a different thread, then this thread will replace
+ *        its pointer to the object with a pointer to the new location.  If the header
+ *        denotes that this object has not yet been copied, this thread will continue to
+ *        use the original cset location as the official version of the object.  Since
+ *        no threads are allowed to allocate for evacuation in this phase, all threads
+ *        accessing this same object will agree to refer to this object at its original
+ *        location within the cset.
+ *     g) Evacuation is cancelled and all threads will eventually reach a Full GC
+ *        safepoint.  Marking by Full GC will finish updating references that might
+ *        be inconsistent within the heap, and will then compact all live memory within
+ *        the heap.
```

Comment:
I like the longer comment you wrote because it provides fuller context and defines the intent of the protocol in greater detail.

I am not sure which I would go with as both descriptions look good to me in their own way. With that in mind, I also left a few comments on the original documentation on the left side.

I am good with whatever you choose to use.